src/cuckoocache.h

```diff
@@ -0,0 +1,457 @@
+        epoch_flags.resize(size);
+        // Set to 45% as described above
+        epoch_size = std::max((uint32_t)1, (45 * size) / 100);
+        // Initially set to wait for a whole epoch
+        epoch_heuristic_counter = epoch_size;
+        return size;
+    }
+
+    /** setup_bytes is a convenience function which accounts for internal memory
+     * usage when deciding how many elements to store. It isn't perfect because
+     * it doesn't account for any overhead (struct size, MallocUsage, collection
+     * and epoch flags). This was done to simplify selecting a power of two
+     * size. In the expected use case, an extra two bits per entry should be
+     * negligible compared to the size of the elements.
+     *
+     * @param bytes the approximate number of bytes to use for this data
+     * structure.
+     * @returns the maximum number of elements storable (see setup()
+     * documentation for more detail)
+     */
+    uint32_t setup_bytes(size_t bytes)
+    {
+        return setup(bytes/sizeof(Element));
+    }
+
+    /** insert loops at most depth_limit times trying to insert a hash
+     * at various locations in the table via a variant of the Cuckoo Algorithm
+     * with eight hash locations.
+     *
+     * It drops the last tried element if it runs out of depth before
+     * encountering an open slot.
+     *
+     * Thus
+     *
+     * insert(x);
+     * return contains(x, false);
+     *
+     * is not guaranteed to return true.
+     *
+     * @param e the element to insert
+     * @post one of the following: All previously inserted elements and e are
+     * now in the table, one previously inserted element is evicted from the
+     * table, the entry attempted to be inserted is evicted.
+     *
+     */
+    inline void insert(Element e)
+    {
+        epoch_check();
+        uint32_t last_loc = invalid();
+        bool last_epoch = true;
+        std::array<uint32_t, 8> locs = compute_hashes(e);
+        // Make sure we have not already inserted this element
+        // If we have, make sure that it does not get deleted
+        for (uint32_t loc : locs)
+            if (table[loc] == e) {
+                please_keep(loc);
+                epoch_flags[loc] = last_epoch;
+                return;
+            }
+        for (uint8_t depth = 0; depth < depth_limit; ++depth) {
+            // First try to insert to an empty slot, if one exists
+            for (uint32_t loc : locs) {
+                if (!collection_flags.bit_is_set(loc))
+                    continue;
+                table[loc] = std::move(e);
+                please_keep(loc);
+                epoch_flags[loc] = last_epoch;
+                return;
+            }
+            /** Swap with the element at the location that was
+            * not the last one looked at. Example:
+            *
+            * 1) On first iter, always false so defaults to locs[0]
+            * 2) Second iter, last_loc == locs[0] so will go to locs[1]
+            *
```

Comment:
I think this comment needs to be updated now that we have more than 2 hashes, right?  If I understand right, the algorithm now is:
```
Swap with the element one past the last one looked at.  Example:

1) On first iter, always false so defaults to locs[0].
2) Second iter, last_loc == locs[k] for some k, so will go to locs[k+1] (wrapping back to 0 if necessary).
```